src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelDiscoveryDataAggregator.cs

```diff
@@ -161,11 +141,93 @@ public void AggregateIsMessageSent(bool isMessageSent)
         IsMessageSent = IsMessageSent || isMessageSent;
     }
 
-    /// <summary>
-    /// Returns sources with particular discovery status.
-    /// </summary>
-    /// <param name="status">Status to filter</param>
-    /// <returns></returns>
-    public List<string> GetSourcesWithStatus(DiscoveryStatus status)
-        => DiscoveryManager.GetSourcesWithStatus(status, _sourcesWithDiscoveryStatus);
+    public List<string> GetSourcesWithStatus(DiscoveryStatus discoveryStatus)
+        => _sourcesWithDiscoveryStatus.IsEmpty
+            ? new List<string>()
+            : _sourcesWithDiscoveryStatus
+                .Where(source => source.Value == discoveryStatus)
+                .Select(source => source.Key)
+                .ToList();
+
+    public void MarkSourcesWithStatus(IEnumerable<string?>? sources, DiscoveryStatus status)
+    {
+        if (sources is null)
+        {
+            return;
+        }
+
+        foreach (var source in sources)
+        {
+            if (source is null)
+            {
+                continue;
+            }
+
+            _sourcesWithDiscoveryStatus.AddOrUpdate(source,
+                _ =>
+                {
+                    if (status != DiscoveryStatus.NotDiscovered)
+                    {
+                        EqtTrace.Warning($"ParallelDiscoveryDataAggregator.MarkSourcesWithStatus: Undiscovered {source}.");
+                    }
+
+                    return status;
+                },
+                (_, previousStatus) =>
+                {
+                    if (previousStatus == DiscoveryStatus.FullyDiscovered && status != DiscoveryStatus.FullyDiscovered
+                        || previousStatus == DiscoveryStatus.PartiallyDiscovered && status == DiscoveryStatus.NotDiscovered)
+                    {
+                        EqtTrace.Warning($"ParallelDiscoveryDataAggregator.MarkSourcesWithStatus: Downgrading source status from {previousStatus} to {status}.");
+                    }
+
+                    EqtTrace.Info($"ParallelDiscoveryDataAggregator.MarkSourcesWithStatus: Marking {source} with {status} status.");
+                    return status;
+                });
+        }
+    }
+
+    public void MarkSourcesBasedOnDiscoveredTestCases(IEnumerable<TestCase>? testCases, bool isComplete, ref string? previousSource)
+    {
+        // When discovery is complete (i.e.not aborted), then we can assume that all partially
+        // discovered sources are now fully discovered. We should also mark all not discovered
+        // sources as fully discovered because that means that the source is considered as tests
+        // but contains no test so we never received the partially or fully discovered event.
+        if (isComplete)
```

Comment:
Is this still true, is this code now sharing all sources in the "cache"? And do we trigger the isComplete from parallel manager or each proxy manager? If we do it from each, don't we mark every source as discovered for every proxy manager that was able to finish?